sdk/sync/client_workers.go

```diff
@@ -0,0 +1,263 @@
+		var del []*Barrier
+		for _, b := range pending {
+			if err := b.ctx.Err(); err != nil {
+				log.Debugw("barrier context expired; removing", "key", b.key)
+				b.C <- err
+				close(b.C)
+				del = append(del, b)
+			}
+		}
+
+		// Prune deleted barriers.
+		for _, b := range del {
+			remove(b)
+		}
+
+		if len(pending) == 0 {
+			// nothing to do; loop over; the ticker will be disarmed.
+			continue
+		}
+
+		// Get the values of all pending states at once, under the context of the Client.
+		log.Debugw("getting barrier values", "keys", keys)
+		vals, err := c.rclient.MGet(keys...).Result()
+		if err != nil {
+			log.Warnw("failed while getting barriers; iteration skipped", "error", err)
+			continue
+		}
+
+		del = del[:0]
+		for i, v := range vals {
+			if v == nil {
+				continue // nobody else has INCR the barrier yet; skip.
+			}
+
+			b := pending[i]
+			curr, err := strconv.ParseInt(v.(string), 10, 64)
+			if err != nil {
+				log.Warnw("failed to parse barrier value", "error", err, "value", v, "key", b.key)
+				continue
+			}
+
+			// Has the barrier been hit?
+			if curr >= b.target {
+				log.Debugw("barrier was hit; informing waiters", "key", b.key, "target", b.target, "curr", curr)
+
+				// barrier has been hit; send a nil error on the channel, and close it.
+				b.C <- nil
+				close(b.C)
+
+				// queue this deletion; otherwise indices won't line up.
+				del = append(del, b)
+			} else {
+				log.Debugw("barrier still unsatisfied", "key", b.key, "target", b.target, "curr", curr)
+			}
+		}
+
+		for _, b := range del {
+			remove(b)
+		}
+	}
+}
+
+func (c *Client) subscriptionWorker() {
+	defer c.wg.Done()
+
+	var (
+		active  = make(map[string]*Subscription)
+		rmSubCh = make(chan []*Subscription, 1)
+	)
+
+	log := c.log.With("process", "subscriptions")
+
+	monitorCtx := func(s *Subscription) {
+		select {
+		case <-s.ctx.Done():
+			log.Debugw("context closure detected; removing subscription", "topic", s.topic.name, "key", s.key)
+			rmSubCh <- []*Subscription{s}
+		case <-c.ctx.Done():
+			log.Debugw("yielding context monitor routine due to global context closure", "topic", s.topic.name, "key", s.key)
+		}
+	}
+
+	consumer := &subscriptionConsumer{c: c, log: log, rmSubCh: rmSubCh, notifyCh: make(chan struct{}, 1)}
+
+	var finalErr error // error to broadcast to remaining subscribers upon returning.
+	defer func() {
+		for _, s := range active {
+			s.doneCh <- finalErr
+			close(s.doneCh)
+			s.outCh.Close()
+		}
+	}()
+
+	for {
+		// Manage subscriptions.
+		select {
+		case newSub := <-c.newSubCh:
+			s := newSub.sub
+			if _, ok := active[s.key]; ok {
+				newSub.resultCh <- fmt.Errorf("failed to add duplicate subscription")
+				continue
+			}
+
+			log.Debugw("adding subscription", "topic", s.topic.name, "key", s.key)
+
+			// interrupt consumer and wait until it yields, before mutating the active set.
+			err := consumer.interrupt()
+			if err != nil {
```

Comment:
We don't seem to be returning an error to `newSub` - wouldn't that deadlock the client?

Reply:
Well spotted!

Reply:
We now panic when this happens, as it is an unrecoverable error: https://github.com/ipfs/testground/pull/841/commits/eabca81cc9fb352aa17475dbd781f0851e4ba3cb